L1Trigger/L1CaloTrigger/plugins/Phase1L1TJetProducer.cc

```diff
@@ -0,0 +1,339 @@
+      inputCollectionTag_{
+          consumes<edm::View<reco::Candidate>>(iConfig.getParameter<edm::InputTag>("inputCollectionTag"))},
+      etaBinning_(iConfig.getParameter<std::vector<double>>("etaBinning")),
+      nBinsEta_(etaBinning_.size() - 1),
+      nBinsPhi_(iConfig.getParameter<unsigned int>("nBinsPhi")),
+      phiLow_(iConfig.getParameter<double>("phiLow")),
+      phiUp_(iConfig.getParameter<double>("phiUp")),
+      jetIEtaSize_(iConfig.getParameter<unsigned int>("jetIEtaSize")),
+      jetIPhiSize_(iConfig.getParameter<unsigned int>("jetIPhiSize")),
+      seedPtThreshold_(iConfig.getParameter<double>("seedPtThreshold")),
+      puSubtraction_(iConfig.getParameter<bool>("puSubtraction")),
+      vetoZeroPt_(iConfig.getParameter<bool>("vetoZeroPt")),
+      outputCollectionName_(iConfig.getParameter<std::string>("outputCollectionName")) {
+  caloGrid_ =
+      std::make_unique<TH2F>("caloGrid", "Calorimeter grid", nBinsEta_, etaBinning_.data(), nBinsPhi_, phiLow_, phiUp_);
+  caloGrid_->GetXaxis()->SetTitle("#eta");
+  caloGrid_->GetYaxis()->SetTitle("#phi");
+  produces<std::vector<reco::CaloJet>>(outputCollectionName_).setBranchAlias(outputCollectionName_);
+}
+
+Phase1L1TJetProducer::~Phase1L1TJetProducer() {}
+
+float Phase1L1TJetProducer::getTowerEnergy(const TH2F& caloGrid, int iEta, int iPhi) {
+  // We return the pt of a certain bin in the calo grid, taking account of the phi periodicity when overflowing (e.g. phi > phiSize), and returning 0 for the eta out of bounds
+
+  int nBinsEta = caloGrid.GetNbinsX();
+  int nBinsPhi = caloGrid.GetNbinsY();
+  while (iPhi < 1) {
+    iPhi += nBinsPhi;
+  }
+  while (iPhi > nBinsPhi) {
+    iPhi -= nBinsPhi;
+  }
+  if (iEta < 1) {
+    return 0;
+  }
+  if (iEta > nBinsEta) {
+    return 0;
+  }
+  return caloGrid.GetBinContent(iEta, iPhi);
+}
+
+void Phase1L1TJetProducer::produce(edm::Event& iEvent, const edm::EventSetup& iSetup) {
+  edm::Handle<edm::View<reco::Candidate>> inputCollectionHandle;
+  iEvent.getByToken(inputCollectionTag_, inputCollectionHandle);
+  // dumping the data
+  caloGrid_->Reset();
+  fillCaloGrid<>(*(caloGrid_), *inputCollectionHandle);
+
+  const auto seedsVector = findSeeds(*(caloGrid_), seedPtThreshold_);  // seedPtThreshold = 6
+  std::vector<reco::CaloJet> l1jetVector;
+  if (puSubtraction_) {
+    l1jetVector = _buildJetsFromSeedsWithPUSubtraction(*(caloGrid_), seedsVector, vetoZeroPt_);
+  } else {
+    l1jetVector = _buildJetsFromSeeds(*(caloGrid_), seedsVector);
+  }
+
+  // sort by pt
+  std::sort(l1jetVector.begin(), l1jetVector.end(), [](const reco::CaloJet& jet1, const reco::CaloJet& jet2) {
+    return jet1.pt() > jet2.pt();
+  });
+
+  std::unique_ptr<std::vector<reco::CaloJet>> l1jetVectorPtr(new std::vector<reco::CaloJet>(l1jetVector));
+  iEvent.put(std::move(l1jetVectorPtr), outputCollectionName_);
+
+  return;
+}
+
+void Phase1L1TJetProducer::subtract9x9Pileup(const TH2F& caloGrid, reco::CaloJet& jet) {
```

Comment:
function can be `const`

Reply:
done: https://github.com/cms-l1t-offline/cmssw/blob/fc3a1a454f40efdf7c95dd97295a6a085a674092/L1Trigger/L1CaloTrigger/plugins/Phase1L1TJetProducer.cc#L168